integrate/romberg.go

```diff
@@ -0,0 +1,98 @@
+// Copyright ©2019 The Gonum Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package integrate
+
+import (
+	"math"
+)
+
+//
+// Helper method for getting the base 2 exponent of an integer
+//
+func getBase2Exponent(x int) int {
+	n := 1
+	k := 0
+
+	for n < x {
+		n <<= 1
+		k += 1
+	}
+	return k
+}
+
+// Romberg estimates the integral of the function f
+// \int_0^1 f(x)dx
+// The function f must be sampled at equal intervals.
+// Romberg's method approximates the integral of the function by
+// applying Richardson extrapolation. Hence, the function adaptively
+// increase the number of evaluation points until the desired accuracy
+// is obtained or the maximum number of evaluation points has been used.
+// See https://en.wikipedia.org/wiki/Romberg%27s_method for a thorough description
+// of the algorithm.
+//
+// The length of the length of the f array must be 2**k + 1, where k is a positive integer
+// The minimum length of the work array is 2*k
+// The method returns the value of the integral, and the number of levels needed to reach
+// the desired accuracy
+func Romberg(f []float64, acc float64, work []float64) (float64, int) {
```

Comment:
Specifying the accuracy tolerance `acc` makes sense for an automatic algorithm so that it can evaluate the integrand only as many times as necessary to reach it. However, here we have _samples_ of the function (the function has already been evaluated) and there is no benefit in not using all of them. Also, once this change is made it makes no sense to return the number of levels used (because all are used).

On the other hand, the function could take the spacing between samples, which would enable arbitrary (finite) integration intervals, although even this parameter can be avoided - the function can assume spacing of 1 but the user then has to rescale the result.

Finally, I'm not sure about the `work` slice. In Gonum we do try to reuse memory but here it spoils what could otherwise be a very clear and simple API:
```
func Romberg(f []float64, dx float64) float64
```
or even
```
func Romberg(f []float64) float64
```
Let's see what others think but I would remove it.

Reply:
I don't think we should include a work slice. AFAIK there are only two places we use it 1) Lapack, but that's different and we have an API around it  2) FFT, and that is also avoidable. 

More generally though, this seems like it's much more suited to `quad` than here. The idea is to be an adaptive algorithm, which is what quad is about.

You could make it a non-adaptive rule (and just evaluate the triangles), but then it's just a special case of the Trapezoidal rule, isn't it? (been a while since I looked at Romberg)

Reply:
I'd be happy for the work slice to go away.